common/src/main/java/com/microsoft/identity/common/internal/eststelemetry/EstsTelemetry.java

```diff
@@ -194,122 +221,139 @@ private IRequestTelemetryCache createLastRequestTelemetryCache(@NonNull final Co
-        lastTelemetry.putTelemetry(Schema.Key.CORRELATION_ID, correlationId);
-        lastTelemetry.putTelemetry(Schema.Key.ERROR_CODE, errorCode);
+        // load the last request object from cache
+        LastRequestTelemetry lastRequestTelemetry = loadLastRequestTelemetryFromCache();
 
-        currentTelemetry.clearTelemetry();
+        // We did not have a last request object in cache, let's create a new one and copySharedValues
+        // fields from current request where applicable
+        if (lastRequestTelemetry == null) {
+            lastRequestTelemetry = new LastRequestTelemetry(currentTelemetry.getSchemaVersion());
+            lastRequestTelemetry = (LastRequestTelemetry) lastRequestTelemetry.copySharedValues(currentTelemetry);
+        }
+
+        if (isTelemetryLoggedByServer(command, commandResult)) {
+            // telemetry headers have been sent to token endpoint and logger by sts
+            // this is the time to reset local telemetry state
+
+            // reset silent successful count as we just went to token endpoint
+            lastRequestTelemetry.resetSilentSuccessCount();
+
+            // get the failed request set for this request. This includes all failed request
+            // data that has been sent to STS in this request.
+            Set<FailedRequest> failedRequestSentSet = mSentFailedRequests.get(correlationId);
+
+            // headers have been logged by sts - we don't need to hold on to this data - let's wipe
+            lastRequestTelemetry.wipeFailedRequestAndErrorForSubList(failedRequestSentSet);
+        }
+
+        // get the error encountered during execution of this command
+        final String errorCode = getErrorFromCommandResult(commandResult);
+
+        if (errorCode != null) {
+            // we have an error, let's append it to the list
+            lastRequestTelemetry.appendFailedRequest(
+                    currentTelemetry.getApiId(),
+                    correlationId,
+                    errorCode);
+        } else if (command instanceof TokenCommand) {
+            final ILocalAuthenticationResult localAuthenticationResult =
+                    (ILocalAuthenticationResult) commandResult.getResult();
+
+            if (localAuthenticationResult.isServicedFromCache()) {
+                // we returned a token from cache, let's increment the silent success count
+                lastRequestTelemetry.incrementSilentSuccessCount();
+            }
+        } // else leave everything as is
+
+        // we're done processing telemetry for this command, let's remove it from the map
         mTelemetryMap.remove(correlationId);
+        mSentFailedRequests.remove(correlationId);
 
-        if (mLastRequestTelemetryCache == null) {
+        if (mLastRequestTelemetryCache != null) {
+            // save the (updated) telemetry object back to telemetry cache
+            mLastRequestTelemetryCache.saveRequestTelemetryToCache(lastRequestTelemetry);
+        } else {
```

Comment:
How would this work for parallel commands?

Each of them will have their own copies,  how are we going to resolve the conflict? As of now I believe the last one wins (which would result in missing telemetries, if any of the commands failed).

Reply:
If it happens at exactly the same time then yes some data will get overriden and some telemetry may go missing. Do you think I should make this method `synchronized`?

Reply:
If we want to make it static, then we have to make flush() static - one has to save before the other loads.

alternatively, we could save the 'changes' instead of the whole request telemetry object. (i.e. delete A, add B).

Reply:
Why static?

Reply:
I was thinking about making both `flush` and `getTelemetryHeaders` synchorized. What do you think about that?

Reply:
That would be simpler to implement, but probably be a bottleneck if there are multiple requests.... so.. tradeoffs here.

I personally prefer #2, but if you think it's too complicated and not worth the performance gain, then I'm also fine with #1 as well. I'll let you make the call.

Reply:
Yeah I think 2 would be complicated to maintain and not worth the effort/performance gain. 

I went ahead and using the 1st approach and addressed it here: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/853/commits/c05a8102c3c4218981eab9abc9ec93d8b4131e79